core/src/worker/activities.rs

```diff
@@ -509,6 +527,53 @@ where
                                     tt,
                                     ActivityCancelReason::WorkerShutdown,
                                 ));
+                            } else {
+                                // Fire off task to keep track of local timeouts. We do this so that
+                                // activities can still get cleaned up even if the user isn't
+                                // heartbeating.
+                                let local_timeout_buffer = self.local_timeout_buffer;
+                                let sched = task
+                                    .resp
+                                    .schedule_to_close_timeout
+                                    .clone()
+                                    .and_then(|to| {
+                                        // Must filter out zero before any subtraction math here
+                                        if to.seconds == 0 && to.nanos == 0 {
+                                            None
+                                        } else {
+                                            Some(to)
+                                        }
+                                    })
+                                    .zip(task.resp.scheduled_time.clone())
+                                    .and_then(|(to, st)| {
+                                        let now = SystemTime::now();
+                                        let sched_time = SystemTime::try_from(st).unwrap_or(now);
+                                        let already_elapsed =
+                                            now.duration_since(sched_time).unwrap_or_default();
+                                        Duration::try_from(to).map(|to| to - already_elapsed).ok()
+                                    });
+                                let timeout_at = [
+                                    task.resp.heartbeat_timeout.clone(),
```

Comment:
You don't want heartbeat timeout here since you don't extend the timeout on heartbeat.
I'd just use the minimum of start to close and schedule to close (which I think the server already sets one of those fields as the minimum, have to verify).

Reply:
Oh, that's a good point. Will change.